src/WebJobs.Script.WebHost/FileMonitoringService.cs

```diff
@@ -98,6 +99,41 @@ public Task StopAsync(CancellationToken cancellationToken)
             return Task.CompletedTask;
         }
 
+        public IDisposable SuspendRestart()
+        {
+            return new SuspendRestartRequest(this);
+        }
+
+        public void ResumeRestart()
+        {
+            Interlocked.Decrement(ref _suspensionRequestsCount);
+            _typedLogger.LogDebug($"Resuming restart. ({_suspensionRequestsCount} requests).");
+        }
+
+        public async Task ScheduleRestartAsync(bool shutdown)
+        {
+            if (shutdown)
+            {
+                _shutdownScheduled = true;
+            }
+
+            if (Interlocked.Read(ref _suspensionRequestsCount) > 0)
+            {
+                _logger.LogDebug("Restart requested while currently suspended. Ignoring request.");
+            }
+            else
+            {
+                if (shutdown || _shutdownScheduled)
```

Comment:
Why is the logic being updated to trigger a shutdown if `_shutdownScheduled` is true? If anything, we should avoid duplicate calls here, correct?

Reply:
Shutdown scheduled should essentially short circuit all of this.

Reply:
this is for scenario:
1. SuspendRestart (shutdown)
2. SuspenRestart (restart)
3. ResumeRestart (shutdown)
4. ResumeRestart (restart)

Shutdown must be performed after step 4.